src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -836,8 +835,8 @@ def create_x509_certificate(self, builder, private_key, algorithm):
         return _Certificate(self, x509_cert)
 
     def _set_asn1_time(self, asn1_time, time):
-        timestamp = calendar.timegm(time.timetuple())
-        res = self._lib.ASN1_TIME_set(asn1_time, timestamp)
+        timestamp = time.strftime('%Y%m%d%H%M%SZ').encode('ascii')
```

Comment:
Unfortunately this incorrectly encodes many values. RFC 5280 requires that times be encoded as UTCTime prior to 2050 and GeneralizedTime thereafter (https://tools.ietf.org/html/rfc5280#section-4.1.2.5). `ASN1_TIME_set_string_X509` does the right thing, but was only added in OpenSSL 1.1.1. If we want to change this we'll need to either replicate the logic in Python (along with plenty of tests to confirm that we're encoding the ASN.1 we expect) or create `Cryptography_ASN1_TIME_set_string_X509` which can duplicate the logic from OpenSSL 1.1.1.

Reply:
```
    /*
     * Per RFC 5280 (section 4.1.2.5.), the valid input time
     * strings should be encoded with the following rules:
     *
     * 1. UTC: YYMMDDHHMMSSZ, if YY < 50 (20YY) --> UTC: YYMMDDHHMMSSZ
     * 2. UTC: YYMMDDHHMMSSZ, if YY >= 50 (19YY) --> UTC: YYMMDDHHMMSSZ
     * 3. G'd: YYYYMMDDHHMMSSZ, if YYYY >= 2050 --> G'd: YYYYMMDDHHMMSSZ
     * 4. G'd: YYYYMMDDHHMMSSZ, if YYYY < 2050 --> UTC: YYMMDDHHMMSSZ
     *
     * Only strings of the 4th rule should be reformatted, but since a
     * UTC can only present [1950, 2050), so if the given time string
     * is less than 1950 (e.g. 19230419000000Z), we do nothing...
     */
```

comment taken from the openssl code

Reply:
Yes, as you can see in that message any time of form `YYYYMMDDHHMMSSZ` must be reformatted to `YYMMDDHHMMSSZ` if `YYYY` < 2050. The PR here will unconditionally set it to `YYYYMMDDHHMMSSZ` because `ASN1_TIME_set_string` doesn't reformat (and the function that does reformat is not available in all supported versions of OpenSSL).

I think implementing this logic in Python and calling `ASN1_TIME_set_string` is a reasonable path forward, but tests will need to be added to check every case + verify that the resulting ASN.1 generated has the correct underlying string type (generalized vs utc).